python/ray/serve/tests/test_deployment_scheduler.py

```diff
@@ -0,0 +1,263 @@
+import sys
+
+import pytest
+
+import ray
+from ray.tests.conftest import *  # noqa
+from ray.serve._private.deployment_scheduler import (
+    DeploymentScheduler,
+    SpreadDeploymentSchedulingPolicy,
+    DriverDeploymentSchedulingPolicy,
+    ReplicaSchedulingRequest,
+    DeploymentDownscaleRequest,
+)
+
+
+@ray.remote(num_cpus=1)
+class Replica:
+    def get_node_id(self):
+        return ray.get_runtime_context().get_node_id()
+
+
+def test_spread_deployment_scheduling_policy_upscale(ray_start_cluster):
+    cluster = ray_start_cluster
+    cluster.add_node(num_cpus=3)
+    cluster.add_node(num_cpus=3)
+    cluster.wait_for_nodes()
+    ray.init(address=cluster.address)
+
+    scheduler = DeploymentScheduler()
+    scheduler.on_deployment_created("deployment1", SpreadDeploymentSchedulingPolicy())
+    replica_actor_handles = []
+    deployment_to_replicas_to_stop = scheduler.schedule(
+        upscales={
+            "deployment1": [
+                ReplicaSchedulingRequest(
+                    deployment_name="deployment1",
+                    replica_name="replica1",
+                    actor_def=Replica,
+                    actor_resources={"CPU": 1},
+                    actor_options={},
+                    actor_init_args=(),
+                    on_scheduled=lambda actor_handle: replica_actor_handles.append(
+                        actor_handle
+                    ),
+                ),
+                ReplicaSchedulingRequest(
+                    deployment_name="deployment1",
+                    replica_name="replica2",
+                    actor_def=Replica,
+                    actor_resources={"CPU": 1},
+                    actor_options={},
+                    actor_init_args=(),
+                    on_scheduled=lambda actor_handle: replica_actor_handles.append(
+                        actor_handle
+                    ),
+                ),
+            ]
+        },
+        downscales={},
+    )
+    assert not deployment_to_replicas_to_stop
+    assert len(replica_actor_handles) == 2
+    assert not scheduler._pending_replicas["deployment1"]
+    assert len(scheduler._launching_replicas["deployment1"]) == 2
+    assert (
+        len(
+            {
+                ray.get(replica_actor_handles[0].get_node_id.remote()),
+                ray.get(replica_actor_handles[1].get_node_id.remote()),
+            }
+        )
+        == 2
+    )
+    scheduler.on_replica_stopping("deployment1", "replica1")
+    scheduler.on_replica_stopping("deployment1", "replica2")
+    scheduler.on_deployment_deleted("deployment1")
+
+
+def test_spread_deployment_scheduling_policy_downscale(ray_start_cluster):
```

Comment:
could you please add short docstrings for these tests describing the specific behavior they're testing? helps *a lot* with readability when others try to modify in the future